### ISSUE TITLE
[Bug]: `data/error1.bal` and `data/hard.bal` tests in semtype port give SO error

### ISSUE THREAD

**Issue by lochana-chathura** [ballerina-platform/ballerina-lang#42662]
### Description

$subject.

Related TODO: https://github.com/ballerina-platform/ballerina-lang/blob/23f1db2da42b4d1a7121c721daa1c73825605922/semtypes/src/main/java/io/ballerina/types/typeops/MappingOps.java#L158-L160
data/error1.bal

<details>
  <summary><b>stacktrace</b>  (<i>click to expand</i>)</summary>

```
java.lang.StackOverflowError
	at io.ballerina.types.subtypedata.BddNode.equals(BddNode.java:32)
	at java.base/java.util.Objects.equals(Objects.java:64)
	at io.ballerina.types.subtypedata.BddNode.equals(BddNode.java:32)
	at java.base/java.util.HashMap.putVal(HashMap.java:633)
	at java.base/java.util.HashMap.put(HashMap.java:610)
	at io.ballerina.types.Common.memoSubtypeIsEmpty(Common.java:200)
	at io.ballerina.types.typeops.ErrorOps.errorSubtypeIsEmpty(ErrorOps.java:50)
	at io.ballerina.types.typeops.ErrorOps.isEmpty(ErrorOps.java:64)
	at io.ballerina.types.Core.isEmpty(Core.java:337)
	at io.ballerina.types.typeops.CellOps.cellMutNoneInhabited(CellOps.java:88)
	at io.ballerina.types.typeops.CellOps.cellInhabited(CellOps.java:78)
	at io.ballerina.types.typeops.CellOps.cellFormulaIsEmpty(CellOps.java:69)
	at io.ballerina.types.Common.bddEvery(Common.java:55)
	at io.ballerina.types.Common.bddEvery(Common.java:58)
	at io.ballerina.types.Common.bddEvery(Common.java:60)
	at io.ballerina.types.typeops.CellOps.cellFormulaIsEmpty(CellOps.java:54)
	at io.ballerina.types.typeops.CellOps.isEmpty(CellOps.java:202)
	at io.ballerina.types.Core.isEmpty(Core.java:337)
	at io.ballerina.types.typeops.MappingOps.mappingInhabited(MappingOps.java:104)
	at io.ballerina.types.typeops.MappingOps.mappingFormulaIsEmpty(MappingOps.java:87)
	at io.ballerina.types.Common.bddEveryPositive(Common.java:67)
	at io.ballerina.types.Common.bddEveryPositive(Common.java:72)
	at io.ballerina.types.Common.bddEveryPositive(Common.java:70)
	at io.ballerina.types.Common.bddEveryPositive(Common.java:70)
	at io.ballerina.types.typeops.ErrorOps.errorBddIsEmpty(ErrorOps.java:54)
	at io.ballerina.types.Common.memoSubtypeIsEmpty(Common.java:205)
	at io.ballerina.types.typeops.ErrorOps.errorSubtypeIsEmpty(ErrorOps.java:50)
	at io.ballerina.types.typeops.ErrorOps.isEmpty(ErrorOps.java:64)
	at io.ballerina.types.Core.isEmpty(Core.java:337)
	at io.ballerina.types.typeops.CellOps.cellMutNoneInhabited(CellOps.java:88)
	at io.ballerina.types.typeops.CellOps.cellInhabited(CellOps.java:78)
	at io.ballerina.types.typeops.CellOps.cellFormulaIsEmpty(CellOps.java:69)
	at io.ballerina.types.Common.bddEvery(Common.java:55)
	at io.ballerina.types.Common.bddEvery(Common.java:58)
	at io.ballerina.types.Common.bddEvery(Common.java:60)
	at io.ballerina.types.typeops.CellOps.cellFormulaIsEmpty(CellOps.java:54)
	at io.ballerina.types.typeops.CellOps.isEmpty(CellOps.java:202)
	at io.ballerina.types.Core.isEmpty(Core.java:337)
	at io.ballerina.types.typeops.MappingOps.mappingInhabited(MappingOps.java:104)
	at io.ballerina.types.typeops.MappingOps.mappingFormulaIsEmpty(MappingOps.java:87)
	at io.ballerina.types.Common.bddEveryPositive(Common.java:67)
	at io.ballerina.types.Common.bddEveryPositive(Common.java:72)
	at io.ballerina.types.Common.bddEveryPositive(Common.java:70)
	at io.ballerina.types.Common.bddEveryPositive(Common.java:70)
	at io.ballerina.types.typeops.ErrorOps.errorBddIsEmpty(ErrorOps.java:54)
	at io.ballerina.types.Common.memoSubtypeIsEmpty(Common.java:205)
	at io.ballerina.types.typeops.ErrorOps.errorSubtypeIsEmpty(ErrorOps.java:50)
	at io.ballerina.types.typeops.ErrorOps.isEmpty(ErrorOps.java:64)
```
</details>

data/hard.bal

<details>
  <summary><b>stacktrace</b>  (<i>click to expand</i>)</summary>

```
java.lang.StackOverflowError
	at io.ballerina.types.Core.diff(Core.java:108)
	at io.ballerina.types.typeops.CellOps.cellMutLimitedInhabited(CellOps.java:107)
	at io.ballerina.types.typeops.CellOps.cellInhabited(CellOps.java:79)
	at io.ballerina.types.typeops.CellOps.cellFormulaIsEmpty(CellOps.java:69)
	at io.ballerina.types.Common.bddEvery(Common.java:55)
	at io.ballerina.types.Common.bddEvery(Common.java:58)
	at io.ballerina.types.Common.bddEvery(Common.java:60)
	at io.ballerina.types.typeops.CellOps.cellFormulaIsEmpty(CellOps.java:54)
	at io.ballerina.types.typeops.CellOps.isEmpty(CellOps.java:202)
	at io.ballerina.types.Core.isEmpty(Core.java:337)
	at io.ballerina.types.typeops.MappingOps.mappingInhabited(MappingOps.java:104)
	at io.ballerina.types.typeops.MappingOps.mappingFormulaIsEmpty(MappingOps.java:87)
	at io.ballerina.types.Common.bddEvery(Common.java:55)
	at io.ballerina.types.Common.bddEvery(Common.java:60)
	at io.ballerina.types.Common.bddEvery(Common.java:58)
	at io.ballerina.types.typeops.MappingOps.lambda$mappingSubtypeIsEmpty$0(MappingOps.java:161)
	at io.ballerina.types.Common.memoSubtypeIsEmpty(Common.java:205)
	at io.ballerina.types.typeops.MappingOps.mappingSubtypeIsEmpty(MappingOps.java:160)
	at io.ballerina.types.typeops.MappingOps.isEmpty(MappingOps.java:237)
	at io.ballerina.types.Core.isEmpty(Core.java:337)
	at io.ballerina.types.typeops.CellOps.cellMutLimitedInhabited(CellOps.java:107)
	at io.ballerina.types.typeops.CellOps.cellInhabited(CellOps.java:79)
	at io.ballerina.types.typeops.CellOps.cellFormulaIsEmpty(CellOps.java:69)
	at io.ballerina.types.Common.bddEvery(Common.java:55)
	at io.ballerina.types.Common.bddEvery(Common.java:58)
	at io.ballerina.types.Common.bddEvery(Common.java:60)
	at io.ballerina.types.typeops.CellOps.cellFormulaIsEmpty(CellOps.java:54)
	at io.ballerina.types.typeops.CellOps.isEmpty(CellOps.java:202)
	at io.ballerina.types.Core.isEmpty(Core.java:337)
	at io.ballerina.types.typeops.MappingOps.mappingInhabited(MappingOps.java:104)
	at io.ballerina.types.typeops.MappingOps.mappingFormulaIsEmpty(MappingOps.java:87)
	at io.ballerina.types.Common.bddEvery(Common.java:55)
	at io.ballerina.types.Common.bddEvery(Common.java:60)
	at io.ballerina.types.Common.bddEvery(Common.java:58)
	at io.ballerina.types.typeops.MappingOps.lambda$mappingSubtypeIsEmpty$0(MappingOps.java:161)
	at io.ballerina.types.Common.memoSubtypeIsEmpty(Common.java:205)
	at io.ballerina.types.typeops.MappingOps.mappingSubtypeIsEmpty(MappingOps.java:160)
	at io.ballerina.types.typeops.MappingOps.isEmpty(MappingOps.java:237)
	at io.ballerina.types.Core.isEmpty(Core.java:337)
	at io.ballerina.types.typeops.CellOps.cellMutLimitedInhabited(CellOps.java:107)
	at io.ballerina.types.typeops.CellOps.cellInhabited(CellOps.java:79)
	at io.ballerina.types.typeops.CellOps.cellFormulaIsEmpty(CellOps.java:69)
	at io.ballerina.types.Common.bddEvery(Common.java:55)
	at io.ballerina.types.Common.bddEvery(Common.java:58)
	at io.ballerina.types.Common.bddEvery(Common.java:60)
	at io.ballerina.types.typeops.CellOps.cellFormulaIsEmpty(CellOps.java:54)
	at io.ballerina.types.typeops.CellOps.isEmpty(CellOps.java:202)
	at io.ballerina.types.Core.isEmpty(Core.java:337)
	at io.ballerina.types.typeops.MappingOps.mappingInhabited(MappingOps.java:104)
	at io.ballerina.types.typeops.MappingOps.mappingFormulaIsEmpty(MappingOps.java:87)
	at io.ballerina.types.Common.bddEvery(Common.java:55)
	at io.ballerina.types.Common.bddEvery(Common.java:60)
	at io.ballerina.types.Common.bddEvery(Common.java:58)
	at io.ballerina.types.typeops.MappingOps.lambda$mappingSubtypeIsEmpty$0(MappingOps.java:161)
	at io.ballerina.types.Common.memoSubtypeIsEmpty(Common.java:205)
	at io.ballerina.types.typeops.MappingOps.mappingSubtypeIsEmpty(MappingOps.java:160)
	at io.ballerina.types.typeops.MappingOps.isEmpty(MappingOps.java:237)
	at io.ballerina.types.Core.isEmpty(Core.java:337)
	at io.ballerina.types.typeops.CellOps.cellMutLimitedInhabited(CellOps.java:107)
	at io.ballerina.types.typeops.CellOps.cellInhabited(CellOps.java:79)
	at io.ballerina.types.typeops.CellOps.cellFormulaIsEmpty(CellOps.java:69)
	at io.ballerina.types.Common.bddEvery(Common.java:55)
	at io.ballerina.types.Common.bddEvery(Common.java:58)
	at io.ballerina.types.Common.bddEvery(Common.java:60)
	at io.ballerina.types.typeops.CellOps.cellFormulaIsEmpty(CellOps.java:54)
	at io.ballerina.types.typeops.CellOps.isEmpty(CellOps.java:202)
	at io.ballerina.types.Core.isEmpty(Core.java:337)
	at io.ballerina.types.typeops.MappingOps.mappingInhabited(MappingOps.java:104)
	at io.ballerina.types.typeops.MappingOps.mappingFormulaIsEmpty(MappingOps.java:87)
	at io.ballerina.types.Common.bddEvery(Common.java:55)
	at io.ballerina.types.Common.bddEvery(Common.java:60)
	at io.ballerina.types.Common.bddEvery(Common.java:58)
	at io.ballerina.types.typeops.MappingOps.lambda$mappingSubtypeIsEmpty$0(MappingOps.java:161)
	at io.ballerina.types.Common.memoSubtypeIsEmpty(Common.java:205)
	at io.ballerina.types.typeops.MappingOps.mappingSubtypeIsEmpty(MappingOps.java:160)
	at io.ballerina.types.typeops.MappingOps.isEmpty(MappingOps.java:237)
```
</details>

**Comment by lochana-chathura** [comment]
Narrowed down to 
```bal
// J1<:J2
// J2<:J1
type J1 record {|J1 x; J1 y;|}|();
type J2 record {|J2 x; J1 y;|}|();
```

However, 
```bal
// J1<:J2
// J2<:J1
type J1 [J2,J1]|();
type J2 [J1,J2]|();
```
works.

**Comment by lochana-chathura** [comment]
One unnecessary cell creation was found.

Now that we handle `CellSemType` instance creation within the `createComplexSemType()` itself, at L101 we get a `CellSemType` instance. Therefore we don't need to create another cell with L103.
https://github.com/ballerina-platform/ballerina-lang/blob/23f1db2da42b4d1a7121c721daa1c73825605922/semtypes/src/main/java/io/ballerina/types/typeops/MappingOps.java#L101-L103

However, the above is not a bug. Just a redundant cell creation. Still the sample is falling with a SO.

**Comment by lochana-chathura** [comment]
Bug found to be in this line,
https://github.com/ballerina-platform/ballerina-lang/blob/23f1db2da42b4d1a7121c721daa1c73825605922/semtypes/src/main/java/io/ballerina/types/Common.java#L191

we are using `listMemo` for mappings and functions.

**Comment by lochana-chathura** [comment]
Fixed in #42679
